File: core/python/src/governor_core/rules.py
"""
Pure-Python evaluator for the Governor rule DSL (spec/schemas/rule.v1.json).

The evaluator operates on plain dicts so it has zero dependency on any
storage layer. Callers translate their own attestation records into the
shape:

    {"actor_id": str, "actor_kind": str}

and pre-compute the actor -> roles mapping:

    {actor_id: {role_slug, ...}}

`evaluate` returns a bool. `explain` returns a recursive structured
reason tree, suitable for serialising into a gate-decision payload.
"""
from __future__ import annotations

from typing import Any, Iterable, Mapping


Attestation = Mapping[str, Any]
ActorRoles = Mapping[str, set]


def _attester_ids(attestations: Iterable[Attestation]) -> set:
    return {str(a['actor_id']) for a in attestations}


def _attester_kinds(attestations: Iterable[Attestation]) -> set:
    return {str(a['actor_kind']) for a in attestations}


def evaluate(
    rule: dict | None,
    attestations: Iterable[Attestation],
    actor_roles: ActorRoles,
) -> bool:
    atts = list(attestations)
    if not isinstance(rule, dict) or not rule:
        return False

    if 'all_of' in rule:
        return all(evaluate(r, atts, actor_roles) for r in rule['all_of'])
    if 'any_of' in rule:
        return any(evaluate(r, atts, actor_roles) for r in rule['any_of'])
    if 'n_of' in rule:
        spec = rule['n_of']
        count = int(spec.get('count', 0))
        branches = spec.get('of', [])
        return sum(1 for r in branches if evaluate(r, atts, actor_roles)) >= count

    if rule.get('actor') is True:
        return len(atts) > 0
    if 'actor_with_role' in rule:
        slug = rule['actor_with_role']
        return any(slug in actor_roles.get(aid, set()) for aid in _attester_ids(atts))
    if 'actor_is' in rule:
        return str(rule['actor_is']) in _attester_ids(atts)
    if 'actor_with_kind' in rule:
        return rule['actor_with_kind'] in _attester_kinds(atts)
    return False


def explain(
    rule: dict | None,
    attestations: Iterable[Attestation],
    actor_roles: ActorRoles,
) -> dict:
    atts = list(attestations)
    node: dict = {'rule': rule, 'satisfied': evaluate(rule, atts, actor_roles)}
    if not isinstance(rule, dict):
        return node
    if 'all_of' in rule:
        node['children'] = [explain(r, atts, actor_roles) for r in rule['all_of']]
    elif 'any_of' in rule:
        node['children'] = [explain(r, atts, actor_roles) for r in rule['any_of']]
    elif 'n_of' in rule:
        node['children'] = [
            explain(r, atts, actor_roles) for r in rule['n_of'].get('of', [])
        ]
    return node

File: core/ts/src/rules.ts
/**
 * Pure-TypeScript evaluator for the Governor rule DSL
 * (spec/schemas/rule.v1.json).
 *
 * Mirrors the Python reference implementation at
 * governor/core/python/src/governor_core/rules.py. Both implementations
 * are validated against the same conformance vectors in
 * governor/spec/conformance/rules.v1.json.
 */

export interface Attestation {
  actor_id: string;
  actor_kind: string;
  [key: string]: unknown;
}

/** Map of actor_id -> set of role slugs that actor holds. */
export type ActorRoles = Record<string, ReadonlySet<string> | readonly string[]>;

export type Rule =
  | null
  | { actor: true }
  | { actor_with_role: string }
  | { actor_with_kind: string }
  | { actor_is: string }
  | { all_of: Rule[] }
  | { any_of: Rule[] }
  | { n_of: { count: number; of: Rule[] } }
  | Record<string, unknown>;

export interface ExplainNode {
  rule: Rule;
  satisfied: boolean;
  children?: ExplainNode[];
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function attesterIds(attestations: readonly Attestation[]): Set<string> {
  return new Set(attestations.map((a) => String(a.actor_id)));
}

function attesterKinds(attestations: readonly Attestation[]): Set<string> {
  return new Set(attestations.map((a) => String(a.actor_kind)));
}

function roleSet(roles: ActorRoles, actorId: string): ReadonlySet<string> {
  const entry = roles[actorId];
  if (!entry) return new Set();
  return entry instanceof Set ? entry : new Set(entry);
}

export function evaluate(
  rule: Rule,
  attestations: readonly Attestation[],
  actorRoles: ActorRoles,
): boolean {
  if (!isObject(rule) || Object.keys(rule).length === 0) {
    return false;
  }

  if ('all_of' in rule) {
    const branches = rule.all_of as Rule[];
    return branches.every((r) => evaluate(r, attestations, actorRoles));
  }
  if ('any_of' in rule) {
    const branches = rule.any_of as Rule[];
    return branches.some((r) => evaluate(r, attestations, actorRoles));
  }
  if ('n_of' in rule) {
    const spec = rule.n_of as { count?: number; of?: Rule[] };
    const count = Number(spec.count ?? 0);
    const branches = spec.of ?? [];
    const hits = branches.reduce(
      (n, r) => n + (evaluate(r, attestations, actorRoles) ? 1 : 0),
      0,
    );
    return hits >= count;
  }

  if ('actor' in rule && rule.actor === true) {
    return attestations.length > 0;
  }
  if ('actor_with_role' in rule) {
    const slug = String(rule.actor_with_role);
    for (const aid of attesterIds(attestations)) {
      if (roleSet(actorRoles, aid).has(slug)) return true;
    }
    return false;
  }
  if ('actor_is' in rule) {
    return attesterIds(attestations).has(String(rule.actor_is));
  }
  if ('actor_with_kind' in rule) {
    return attesterKinds(attestations).has(String(rule.actor_with_kind));
  }
  return false;
}

export function explain(
  rule: Rule,
  attestations: readonly Attestation[],
  actorRoles: ActorRoles,
): ExplainNode {
  const node: ExplainNode = {
    rule,
    satisfied: evaluate(rule, attestations, actorRoles),
  };
  if (!isObject(rule)) return node;
  if ('all_of' in rule) {
    node.children = (rule.all_of as Rule[]).map((r) => explain(r, attestations, actorRoles));
  } else if ('any_of' in rule) {
    node.children = (rule.any_of as Rule[]).map((r) => explain(r, attestations, actorRoles));
  } else if ('n_of' in rule) {
    const branches = (rule.n_of as { of?: Rule[] }).of ?? [];
    node.children = branches.map((r) => explain(r, attestations, actorRoles));
  }
  return node;
}
